lib/model/indexsorter.go

```diff
@@ -0,0 +1,224 @@
+	if is, ok := s.indexSorter.(*inMemoryIndexSorter); ok {
+		// We have an in memory index sorter. Check if the size has been
+		// exceeded, and if so switch to a database backed index sorter before
+		// continuing.
+
+		if is.size > s.maxInMemorySize {
+			ds, err := newLeveldbIndexSorter()
+			if err != nil {
+				panic(err) // What else, at this point?
+			}
+			ds.Enqueue(is.items...)
+			s.indexSorter = ds
+		}
+	}
+
+	s.indexSorter.Enqueue(items...)
+}
+
+// An inMemoryIndexSort is simply a []protocol.FileInfo that makes sure to
+// sort itself on LocalVersion when the user starts asking for batches.
+type inMemoryIndexSorter struct {
+	items   []protocol.FileInfo
+	sorted  bool
+	nextIdx int
+	size    int
+}
+
+func (s *inMemoryIndexSorter) Enqueue(items ...protocol.FileInfo) {
+	if s.sorted {
+		panic("bug: enqueue on sorted indexSorter")
+	}
+
+	// Append the files to our slice, and increase the size by how much space
+	// we think this consumes.
+
+	s.items = append(s.items, items...)
+	for _, f := range items {
+		s.size += sizeBytes(f)
+	}
+}
+
+func (s *inMemoryIndexSorter) Batch() []protocol.FileInfo {
+	if !s.sorted {
+		sort.Sort(sortByLocalVersion(s.items))
+		s.sorted = true
+	}
+
+	if s.nextIdx >= len(s.items) {
+		// nextIdx is already at the end, nothing more to return.
+		return nil
+	}
+
+	// Find the end of the current batch; it needs to be smaller than
+	// batchTargetSize bytes, and fewer than batchMaxFiles, and not mroe than
+	// we actually have in the slice.
+
+	end := s.nextIdx
+	batchSize := 0
+	for end < len(s.items) && batchSize < batchTargetSize && end-s.nextIdx < batchMaxFiles {
+		batchSize += sizeBytes(s.items[end])
+		end++
+	}
+
+	batch := s.items[s.nextIdx:end]
+	s.nextIdx = end
+
+	return batch
+}
+
+func (s *inMemoryIndexSorter) Close() {
+	// Nothing to clean up here.
+}
+
+// A leveldbIndexSorter inserts items keyed on LocalVersion, then iterates
+// over that when asked for batches.
+type leveldbIndexSorter struct {
+	db      *leveldb.DB
+	dbPath  string
+	prevKey int64
+}
+
+func newLeveldbIndexSorter() (*leveldbIndexSorter, error) {
+	path, err := ioutil.TempDir("", "indexCache")
+	if err != nil {
+		return nil, err
+	}
+	db, err := leveldb.OpenFile(path, &opt.Options{
+		// We're only going to read it once, so we don't need caching
+		DisableBlockCache: true,
+		// It should certainly not already exist
+		ErrorIfExist: true,
+		// We don't need sync writes on it, the data is worthless if we crash
+		NoSync: true,
+		// And we don't want to spend too many open files on it
+		OpenFilesCacheCapacity: 16,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	l.Debugln("leveldbIndexSorter at", path)
+	return &leveldbIndexSorter{
+		db:     db,
+		dbPath: path,
+	}, nil
+}
+
+func (s *leveldbIndexSorter) Enqueue(items ...protocol.FileInfo) {
+	// Marshal each item, keyed by it's LocalVersion.
+	for _, f := range items {
+		var key [8]byte
+		binary.BigEndian.PutUint64(key[:], uint64(f.LocalVersion))
```

Comment:
As of [this](https://github.com/syncthing/syncthing/pull/2020/files#diff-69b15624219bb5333982e6cf1df4ef78R32) it's globally unique, however there may be an issue here if the device's clock is set backwards and Syncthing restarted...